### PR TITLE
Experimental: Strictly typed reactive forms 

### PR DESCRIPTION
**Pull request by sonukapoor**
Based on the original work from #37389

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [X] Feature

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:325946371:BASE:325946388:1597117795227:f501461d).

File: goldens/public-api/forms/forms.d.ts [hunk 1]
```diff
@@ -205,37 +207,39 @@ export declare class FormArrayName extends ControlContainer implements OnInit, O
...
         [key: string]: any;
     } | null): FormGroup;
 }
 
-export declare class FormControl extends AbstractControl {
-    constructor(formState?: any, validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null);
-    patchValue(value: any, options?: {
+export declare class FormControl<T = any> extends AbstractControl<T> {
+    constructor(formState?: FormControlState<T>, validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null);
+    patchValue(value: null | T, options?: {
```
Comment (destus90) [inline goldens/public-api/forms/forms.d.ts:221]: Is it correct? As described in the docs, we might provide some fields we would like to update, not the full model structure. With these changes, we will have some TypeScript types errors.
```suggestion
    patchValue(value: null | Partial<T>, options?: {
```

Comment (KostyaTretyak) [reply goldens/public-api/forms/forms.d.ts:221]: @destus90 , I think it's correct, `Partial<T>` need only for `FormGroup`. Please give an example with `FormControl` where `Partial<T>` really needs to be used.

Comment (destus90) [reply goldens/public-api/forms/forms.d.ts:221]: Ok, you are right. At first I thought these changes will affect to all the reactive directives.

**Comment by AndrewKushnir** [comment]
Hi,

Quick update after running tests in Google's codebase: this change is breaking a portion of targets.

I performed initial investigation and it looks like a lot of targets are failing with a similar error and we also have this scenario in tests:

```
const c = new FormControl(null, {validators});
c.setValue('value');
```

TypeScript infers the generic type (`T`) as `null`, thus the `setValue` call argument with the string type produces an error. In the tests this is solved by adding the `<string>` generic type, i.e.:

```
const c = new FormControl<string>(null, {validators});
c.setValue('value');
```

It'd be great to investigate if adding a generic type can be avoided in this case (without loosing type safety) - it should reduce the number of failing targets.

Thank you.

**Comment by JoostK** [comment]
Quick note on the current state of the PR: Sonu and I made some changes over the weekend to strongly type `AbstractControl.value` and modified some constructor overloads in order to allow type inference to work as expected, but then hit a bit of a problem that is causing tests to fail left and right.

The gist of the issue that is yet to be resolved is how to deal with nullability in form controls, where type inference did not include the null. For example:

```ts
const group = new FormGroup({ name: 'Joost', age: 28 });
```

This will be inferred as `FormGroup<{name: string, age: number}>`, so `FormGroup.value` is typed as `{name: string, age: number}|null`, whereas it could be `{name: string|null, age: number|null}`. We played with a mapped type in `FormGroup.value`'s type, however that will start generating an incompatible redeclaration of `AbstractControl.value` as it is `T|null`.

---

Another use-case that needs some discussion is how to deal with `addControl`. This is an interesting API in terms of the keys that can be added, and how that newly added key can be accessed. If the `FormGroup` models some interface, then you'd expect the keys you can add to be restricted to the properties declared in the interface, but if the type is inferred then you may want to add additional controls after construction. The latter is problematic, given that it can't reflect the newly added key in te generic type (that would require `addControl` to have a return type and reassignment of the whole group). Also, the API can't differentiate between explicitly declared generic types (where it makes sense for `addControl` to be restricted) vs. an inferred type based on how the `FormGroup` was created (but no generic type declared). Add to that the backwards compatibility concert and you have several more constraints.

**Comment by AndrewKushnir** [comment]
[Presubmit #2](http://test/OCL:334386792:BASE:334386832:1601394323682:a1a6fcad).

File: goldens/public-api/forms/forms.d.ts [hunk 2]
```diff
@@ -31,7 +31,7 @@ export declare abstract class AbstractControl {
         onlySelf?: boolean;
         emitEvent?: boolean;
     }): void;
-    get(path: Array<string | number> | string): AbstractControl | null;
+    get<T = any>(path: Array<string | number> | string): AbstractControl<T> | null;
```
Comment (wiesnery) [inline goldens/public-api/forms/forms.d.ts:34]: can't we use something like:
`get<S extends keyof T>(path: Array<string | number> | S): AbstractControl<T[S]> | null  {`

This way we can achieve that we can only access FormControls wich are defined in the type.

File: goldens/public-api/forms/forms.d.ts [hunk 3]
```diff
@@ -18,8 +18,8 @@ export declare abstract class AbstractControl {
     get valid(): boolean;
     get validator(): ValidatorFn | null;
     set validator(validatorFn: ValidatorFn | null);
-    readonly value: any;
-    readonly valueChanges: Observable<any>;
+    readonly value: T | null;
```
Comment (wiesnery) [inline goldens/public-api/forms/forms.d.ts:21]: Maybe we could also think of using `Partial<T> | null` as return type, as some values might have a null value if they weren't initialized and we provide a useful hint to the consumer of the object.

File: goldens/public-api/forms/forms.d.ts [hunk 4]
```diff
@@ -167,31 +167,22 @@ export declare interface Form {
     updateModel(dir: NgControl, value: any): void;
 }
 
-export declare class FormArray extends AbstractControl {
-    controls: AbstractControl[];
-    get length(): number;
-    constructor(controls: AbstractControl[], validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null);
-    at(index: number): AbstractControl;
+export declare interface FormArray<T = any> extends AbstractControl<T> {
```
Comment (leo6104) [inline goldens/public-api/forms/forms.d.ts:170]:
```
export declare interface FormArray<T = any> extends AbstractControl<T[]> {
```
its AbstractControl should have T[] as value type. `AbstractControl<T[]>` 

```
formArray = new FormArray([ new FormControl(1), new FormControl(2) ]);
formArray.value // [ 1, 2 ] 
```

**Review by leo6104 — CHANGES_REQUESTED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:366177614:BASE:366177629:1617249967434:2a0247c).

File: packages/forms/test/value_accessor_integration_spec.ts [hunk 5]
```diff
@@ -64,7 +64,7 @@ import {dispatchEvent} from '@angular/platform-browser/testing/src/browser_util'
 
       const input = fixture.debugElement.query(By.css('input'));
       form.valueChanges.subscribe({
-        next: (value) => {
+        next: (value: any) => {
```
Comment (ydmitry) [inline packages/forms/test/value_accessor_integration_spec.ts:67]: `unknown` instead of `any` could be better here

**Comment by atscott** [comment]
Closing since this is being actively worked on by @dylhunn as part of the typed forms project (#43834, #44513)

**Pull request closed without merging**